state-chain/pallets/cf-environment/src/lib.rs

```diff
@@ -745,6 +816,39 @@ impl<T: Config> Pallet<T> {
 			log::error!("Nonce account {nonce_account} not found in unavailable nonce accounts");
 		}
 	}
+
+	pub fn report_sol_contract_swap_accounts(
+		new_contract_swaps: Vec<(ContractSwapAccountAndSender, SolanaContractSwapDetails)>,
+		confirm_closed_accounts: Vec<ContractSwapAccountAndSender>,
+	) {
+		SolanaOpenContractSwapAccounts::<T>::mutate(|accts| {
+			accts.extend(
+				new_contract_swaps
+					.into_iter()
+					.map(|(contract_swap_account, swap_details)| {
+						T::SolanaContractSwapper::initiate_contract_swap(
+							swap_details.from,
+							swap_details.to,
+							swap_details.deposit_amount,
+							swap_details.destination_address,
+							swap_details.tx_hash,
+						);
+						contract_swap_account
+					})
+					.collect::<Vec<_>>(),
+			)
+		});
+		confirm_closed_accounts
+			.into_iter()
+			.for_each(SolanaClosedContractSwapAccounts::<T>::remove);
```

Comment:
Would it maybe make sense to log something if we have tried to remove an account that doesn't exist? If it happens, unless it's a big reorg, we might want to know.

Reply:
This should not be possible given the way how our witnessing for these accounts would work. The witnessing would get the list of ```SolanaClosedContractSwapAccounts``` from the state chain and compare this list with the witnessed accounts list and submit to the state chain as ```confirm_closed_aacounts``` those accounts that are present in the SolanaClosedContractSwapAccounts but not in the witnessed_accounts list (which means they are successfully closed).

Reply:
True, I was thinking from the point of view of the on-chain program. All good :+1: